src/peft/tuners/vera/layer.py

```diff
@@ -0,0 +1,676 @@
+        self.r[adapter_name] = r
+        if vera_dropout > 0.0:
+            vera_dropout_layer = nn.Dropout(p=vera_dropout)
+        else:
+            vera_dropout_layer = nn.Identity()
+
+        self.vera_dropout[adapter_name] = vera_dropout_layer
+        # Actual trainable parameters
+        if r > 0:
+            kernel_size = self.kwargs["kernel_size"]
+            stride = self.kwargs["stride"]
+            padding = self.kwargs["padding"]
+
+            # TODO: these need to be frozen and initialised correctly somewhere
+            self.vera_A[adapter_name] = nn.Conv2d(self.in_features, r, kernel_size, stride, padding, bias=False)
+            self.vera_B[adapter_name] = nn.Conv2d(r, self.out_features, (1, 1), (1, 1), bias=False)
+
+            self.vera_lambda_b[adapter_name] = nn.Parameter(torch.zeros(self.out_features), requires_grad=True)
+            self.vera_lambda_d[adapter_name] = nn.Parameter(torch.zeros(r), requires_grad=True)
+
+        if init_vera_weights:
+            self.reset_vera_parameters(adapter_name, prng_key, d_initial=d_initial)
+
+        weight = getattr(self, "weight", None)
+        if weight is not None:
+            # the layer is already completely initialized, this is an update
+            self.to(self.weight.device, dtype=weight.dtype)
+
+    def update_layer_embedding(self, adapter_name, r, vera_dropout, init_vera_weights, prng_key, d_initial: float = 1.0):
+        if r <= 0:
+            raise ValueError(f"`r` should be a positive integer value but the value passed is {r}")
+        self.r[adapter_name] = r
+        if vera_dropout > 0.0:
+            vera_dropout_layer = nn.Dropout(p=vera_dropout)
+        else:
+            vera_dropout_layer = nn.Identity()
+
+        self.vera_dropout[adapter_name] = vera_dropout_layer
+        # Actual trainable parameters
+        if r > 0:
+            weight_A = torch.randn((r, self.in_features))
+            weight_B = torch.randn((self.out_features, r))
+
+            # TODO: these need to be frozen and initialised correctly somewhere
+            self.vera_embedding_A[adapter_name] = nn.Parameter(weight_A)
+            self.vera_embedding_B[adapter_name] = nn.Parameter(weight_B)
+
+            self.vera_lambda_b[adapter_name] = nn.Parameter(torch.zeros(self.out_features), requires_grad=True)
+            self.vera_lambda_d[adapter_name] = nn.Parameter(torch.zeros(r), requires_grad=True)
+
+        if init_vera_weights:
+            self.reset_vera_parameters(adapter_name, prng_key, d_initial=d_initial)
+
+        weight = getattr(self, "weight", None)
+        if weight is not None:
+            # the layer is already completely initialized, this is an update
+            self.to(self.weight.device, dtype=weight.dtype)
+
+    def reset_vera_parameters(self, adapter_name, prng_key, d_initial: float = 1.0):
+        generator = torch.Generator(device='cpu').manual_seed(prng_key)
```

Comment:
I'm not 100% sure, but it seems to me that the paper says that the A and B matrices are actually generated on the fly (i.e. during `forward`) based on the seed:

> Because the random frozen matrices can be regenerated from a
random number generator (RNG) seed, these do not need to be stored in memory. This substantially
reduces the memory requirement, which is now limited to the bytes needed for the trained b and d
vectors and a single RNG seed

However, this should be slower than generating them once and storing them. I guess in the end it's a tradeoff between runtime performance and potential peak memory.

Reply:
You are right, I stored for now for experimentation. See my reply to your main comment and we can come to an agreement on whether to support regenerating, storing, or both, then I will implement it.